Add SyncAsyncEventHandler (#18170)
Azure/azure-sdk-for-net: sdk/core/Azure.Core/samples/Events.md

# Azure.Core Event samples

**NOTE:** Samples in this file only apply to packages following the
[Azure SDK Design Guidelines](https://azure.github.io/azure-sdk/dotnet_introduction.html).
The names of these packages usually start with `Azure`.

> Most Azure client libraries offer both synchronous and asynchronous methods for

Reviewer: ```suggestion
Most Azure SDK .NET client libraries offer both synchronous and asynchronous methods for
```

Author: I added "for .NET" to match what we usually write like https://github.com/Azure/azure-sdk-for-net/tree/master/sdk/storage/Azure.Storage.Blobs#azure-storage-blobs-client-library-for-net.

calling Azure services.  You can distinguish the asynchronous methods by their
`Async` suffix.  For example, `BlobClient.Download` and `BlobClient.DownloadAsync`
make the same underlying REST call and only differ in whether they block.  We
recommend using our async methods for new applications, but there are perfectly
valid cases for using sync methods as well.  These dual method invocation

> semantics work great, but require a little extra care when writing event handlers.

Reviewer: ```suggestion
semantics address the needs of our customers but require a little extra care when writing event handlers.
```

Author: Fixed

> The `SyncAsyncEventHandler` is a delegate used by events in Azure client

Reviewer: For event handler and event args ... link to the source or docs so folks can learn more?

Author: There are no docs (yet) and I don't think the source adds any value compared to what's here (i.e., [Handler](https://github.com/Azure/azure-sdk-for-net/blob/f27c837934e9eb68a74aca4e1a5d3fabbe33d40c/sdk/core/Azure.Core/src/SyncAsyncEventHandler.cs) and [Args](https://github.com/Azure/azure-sdk-for-net/blob/f27c837934e9eb68a74aca4e1a5d3fabbe33d40c/sdk/core/Azure.Core/src/SyncAsyncEventArgs.cs)).

libraries to represent an event handler that can be invoked from either sync or
async code paths.  It takes event arguments deriving from `SyncAsyncEventArgs`
that contain important information for writing your event handler.

- `SyncAsyncEventArgs.CancellationToken` is a cancellation token related to the
  original operation that raised the event.  It's important for your handler to
  pass this token along to any asynchronous or long-running synchronous
  operations that take a token so cancellation (via something like
  `new CancellationTokenSource(TimeSpan.FromSeconds(10)).Token`, for example)
  will correctly propagate.

- There is a `SyncAsyncEventArgs.RunSynchronously` flag indicating whether your
  handler was invoked synchronously or asynchronously.  In general,

>     - If you're calling sync methods on your client, you should use sync methods

Reviewer: It's not clear to me if the reader you're addressing here is the library user or library developer.  Would it be possible to clarify?

Author: The whole document is for users of the library.

      to implement your event handler.  You can return `Task.CompletedTask`.
    - If you're calling async methods on your client, you should use async
      methods where possible to implement your event handler.
    - If you're not in control of how the client will be used or want to write
      safer code, you should check the `RunSynchronously` property and call
      either sync or async methods as directed.

  There are code examples of all three situations below to compare.  Please also

>   see the note at the very end discussing the dangers of sync-over-async to

Reviewer: Link to section header?

Author: I haven't seen that in any other Azure.Core samples and it feels a little short for links between sections.

  understand the risks of not using the `RunSynchronously` flag.

- Most events will customize the event data by deriving from `SyncAsyncEventArgs`
  and including details about what triggered the event or providing options to
  react.  Many times this will include a reference to the client that raised the
  event in case you need it for additional processing.

When an event using `SyncAsyncEventHandler` is raised, the handlers will be

> executed sequentially to avoid introducing any unintended parallelism.  The

Reviewer: ```suggestion
executed sequentially to avoid introducing any unintended parallelism to calling code.  The
```

Author: I think "calling code" muddies it a bit since we're avoiding unintentional parallelism in the handlers.

event handlers will finish before returning control to the code path raising the
event.  This means blocking for events raised synchronously and waiting for the
returned `Task` to complete for events raised asynchronously.  Any exceptions

> thrown from a handler will be wrapped in a single `AggregateException`.  Finally,

Reviewer: Link here to the section that describes how exceptions work here?

Author: I haven't seen that in any other Azure.Core samples and it feels a little short for links between sections.

> we wrap a [distributed tracing span](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/core/Azure.Core/samples/Diagnostics.md#distributed-tracing)

Reviewer: Who is we?  Does this happen automatically in Azure.Core?  Would be good to include a few words or sentence to clarify.

Author: It happens automatically using the shared `RaiseAsync` extension method.

returned `Task` to complete for events raised asynchronously.  Any exceptions
thrown from a handler will be wrapped in a single `AggregateException`.  Finally,
a [distributed tracing span](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/core/Azure.Core/samples/Diagnostics.md#distributed-tracing)
is wrapped around your handlers using the event name so you can see how long
your handlers took to run, whether they made other calls to Azure services, and
details about any exceptions that were thrown.

The rest of the code samples are using a fictitious `AlarmClient` to demonstrate
how to handle `SyncAsyncEventHandler` events.  There are `Snooze` and
`SnoozeAsync` methods that both raise a `Ring` event.

## Adding a synchronous event handler

If you're using the synchronous, blocking methods of a client (i.e., methods
without an `Async` suffix), they will raise events that require handlers to
execute synchronously as well.  Even though the signature of your handler
returns a `Task`, you should write regular sync code that blocks and return
`Task.CompletedTask` when finished.

```C# Snippet:Azure_Core_Samples_EventSamples_SyncHandler
var client = new AlarmClient();

> client.Ring += (SyncAsyncEventArgs e) =>

Reviewer 2: Need an example with unsubscribing from event.

Author: I don't think there's a good reason to.  There's nothing special about our model that changes unsubscribing and it's not a thing we'd expect most customers to do.
